src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -544,18 +545,28 @@ public byte[] Http2AltSvcOriginUri
                         Trace("Attempting new HTTP2 connection.");
                     }
 
-                    Stream? stream;
                     HttpResponseMessage? failureResponse;
-                    (socket, stream, transportContext, failureResponse) =
+
+                    (connection, transportContext, failureResponse) =
                         await ConnectAsync(request, async, true, cancellationToken).ConfigureAwait(false);
+
                     if (failureResponse != null)
                     {
                         return (null, true, failureResponse);
                     }
 
+                    Debug.Assert(connection != null);
+
+                    sslStream = (SslStream)connection.Stream;
```

Comment:
I haven't looked at the context around this line closely, but have you tested your new logic with HTTP/2 + H2C? I'm guessing the stream won't be SslStream in that case.

Reply:
Tests pass but I don't recall what tests we have for H2C. I'll double check that this is correct, thanks.

Reply:
It seems we have no tests for H2C :). Thanks for pointing this out, fixed.

We'll want to add some tests for H2C, perhaps as part of https://github.com/dotnet/runtime/pull/39201 or elsewhere. I filed [an issue](https://github.com/dotnet/runtime/issues/39648) for it.

Reply:
H2C has some coverage on Mac where server-side ALPN is not available. So I fixed some tests to fall-back to HTTP/2 even without SSL.